### PR TITLE
Shader -> SPIR-V at CTS build time (#1849)

### LINKED ISSUE #1849: Vulkan interop tests should not rely on checked-in SPIR-V files
We ought to compile the shaders at CTS build time or integrate glslang and build them programmatically.

### PR DESCRIPTION
**Pull request by saurabhnv**
Instead of relying on prebuilt checked-in SPIR-V files, compile available shaders at CTS build time.

* Add dependency on glslc (available as part of VULKAN_SDK).
* Add optional build flag BUILD_GLSL_SHADERS, OFF by default.

Fixes #1849

### CODE REVIEW

**Comment by bashbaug** [comment]
Nikhil and I tried to add "focused review" at the same time, which is why the label was added and then immediately removed.  This should have "focused review".

**Comment by nikhiljnv** [comment]
@kpet The way this change is setup currently is that GLSL shaders are compiled to generate SPIR-V files with the same name as existing prebuilt SPIR-V (*.spv) files under https://github.com/KhronosGroup/OpenCL-CTS/tree/main/test_conformance/vulkan/shaders. So, if the option to compile GLSL shaders is enabled during test build, it replaces the prebuilt *.spv files with the ones generated by glslc shader compiler and the same will be used by existing tests in place of prebuilt shaders.

We can remove prebuilt *.spv files to avoid confusion, but prefer to do it in a subsequent PR.

**Comment by bashbaug** [comment]
Feedback from the January 30th and February 6th teleconferences: it would be best to remove the check-in SPIR-V files as part of this change so there is only one valid way for testing.

File: presubmit.sh [hunk 1]
```diff
@@ -4,6 +4,10 @@ set -e
 
 export TOP=$(pwd)
 
+GLSLANG_URL_Windows="https://github.com/KhronosGroup/glslang/releases/download/main-tot/glslang-master-windows-Release.zip"
+GLSLANG_URL_Linux="https://github.com/KhronosGroup/glslang/releases/download/main-tot/glslang-main-linux-Release.zip"
+GLSLANG_URL_macOS="https://github.com/KhronosGroup/glslang/releases/download/main-tot/glslang-main-osx-Release.zip"
+
```
[the remark this answers is not in the document]
Comment (saurabhnv) [reply presubmit.sh:10]: Agree about the exposure to unnecessary noise with ToT build.
We can use a fixed version, either 14.0.0 or recently released one 14.1.0 should work.

As for handling the updates, a fixed version should be fine until and unless we update the shaders to use functionality not supported by that version, so a manual update (in presubmit.sh whenever required) of the version number should be fine, thoughts?

[the remark this answers is not in the document]
Comment (saurabhnv) [reply presubmit.sh:10]: Looks like main-tot is the only release that contains pre-built glslang, otherwise just the source is available.
Should we build it then?

File: test_conformance/vulkan/CMakeLists.txt [hunk 2]
```diff
@@ -43,4 +43,6 @@ set_source_files_properties(
     PROPERTIES LANGUAGE CXX)
 include_directories("../common/vulkan_wrapper")
 
+include(shaders/CMakeLists.txt)
```
[the remark this answers is not in the document]
Comment (saurabhnv) [reply test_conformance/vulkan/CMakeLists.txt]: Addressed in recent commit.

File: presubmit.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpet) [reply presubmit.sh:10]: We should probably use the Vulkan SDK then. You can have a look at https://github.com/kpet/clvk/blob/main/.github/actions/install-vulkan-sdk/action.yml for some inspiration.

File: test_conformance/vulkan/shaders/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    file(READ ${IMAGE2D_SHADER_IN_FILE} IMAGE2D_SHADER_UNFORMAT_CONTENT)
+    file(STRINGS ${IMAGE2D_FORMATS_LIST_IN_FILE} IMAGE2D_FORMATS_LIST)
+
+    foreach(IMAGE2D_FORMAT ${IMAGE2D_FORMATS_LIST})
+        list(GET IMAGE2D_FORMAT 1 GLSL_FORMAT)
+        list(GET IMAGE2D_FORMAT 2 GLSL_TYPE_PREFIX)
+        string(REPLACE "GLSL_FORMAT" "${GLSL_FORMAT}" IMAGE2D_SHADER_CONTENT "${IMAGE2D_SHADER_UNFORMAT_CONTENT}")
+        string(REPLACE "GLSL_TYPE_PREFIX" "${GLSL_TYPE_PREFIX}" IMAGE2D_SHADER_CONTENT "${IMAGE2D_SHADER_CONTENT}")
+        file(WRITE ${IMAGE2D_SHADER_TMP_OUT_FILE} "${IMAGE2D_SHADER_CONTENT}")
+        execute_process(
```
Comment (kpet) [inline test_conformance/vulkan/shaders/CMakeLists.txt:27]: Generating the SPIR-V modules as part of the configuration stage is not great and should ideally be done as separate targets. I'm fine if you want to do this as a separate PR but I think we really should tidy this up.

**Review by kpet — APPROVED** [review]
Ok as a first step.

File: test_conformance/vulkan/shaders/CMakeLists.txt (hunk 3 above)
Comment (saurabhnv) [reply test_conformance/vulkan/shaders/CMakeLists.txt:27]: Sure, will raise one

**Comment by bashbaug** [comment]
Merging as discussed in the April 2nd memory subgroup.
